src/security/Crypto.js

```diff
@@ -1,135 +1,183 @@
-import nacl from "tweetnacl";
-import sha256 from "fast-sha256";
 import base64Url from "base64url";
 import stringify from "json-stable-stringify";
-import Util from "../Util";
 import {Buffer} from "buffer/.";
+import nacl from "tweetnacl";
+import sha256 from "fast-sha256";
 
-let sjcl = null;
-
-/**
- * Initializes sjcl and mouse input collection, for IE10
- */
-if (!BROWSER) {
-} else if (window.crypto && window.crypto.getRandomValues) {
-  // Do nothing, we have secure random crypto
-} else if (window.msCrypto && window.msCrypto.getRandomValues) {
-  // Do nothingm, we have secure random crypto
-} else {
-    // Only set it up when it is necessary
-    sjcl = require('sjcl');
-    sjcl.random.startCollectors();
+if (BROWSER && !crypto) {
+    throw new Error('Your browser does not support Web Cryptography API');
 }
 
 /**
  * Class providing static crypto primitives.
  */
 class Crypto {
     /**
-     * Generates a keypair to use with the token System
+     * Generates a key pair to use with the Token system.
      *
-     * @param {string} keyLevel - desired security level of key
-     * @return {object} keyPair - keyPair
+     * @param {string} keyLevel - "LOW", "STANDARD", or "PRIVILEGED"
+     * @param {boolean} extractable - whether the private key can be extracted from the CryptoKey object
+     * @return {Object} generated key pair
      */
-    static generateKeys(keyLevel) {
-        if (sjcl !== null) {
-            if (sjcl.random.isReady()) {
-                nacl.setPRNG(function(x, n) {
-                    const randomWords = sjcl.random.randomWords(n / 4);
-                    for (let i = 0; i < n; i++) {
-                        x[i] = Util.getByte(randomWords[Math.floor(i / 4)], i % 4);
-                    }
-                    return x;
-                });
-            } else {
-                throw new Error('Not enough entropy for random number generation');
-            }
+    static async generateKeys(keyLevel, extractable = false) {
+        if (!BROWSER) {
+            const keyPair = nacl.sign.keyPair();
+            keyPair.id = base64Url(sha256(keyPair.publicKey)).substring(0, 16);
+            keyPair.algorithm = 'ED25519';
+            keyPair.level = keyLevel;
+            keyPair.privateKey = keyPair.secretKey;
+            delete keyPair.secretKey;
+            return keyPair;
+        }
+        const keyPair = await crypto.subtle.generateKey(
+            {
+                name: 'ECDSA',
+                namedCurve: 'P-256',
+            },
+            extractable,
+            ['sign', 'verify'],
+        );
+        keyPair.publicKey = new Uint8Array(await crypto.subtle.exportKey('spki', keyPair.publicKey));
+        if (extractable) {
+            keyPair.privateKey = new Uint8Array(await crypto.subtle.exportKey('pkcs8', keyPair.privateKey));
```

Comment:
why do we need to allow it to be extractable?

Reply:
for UnsecuredFileCryptoEngine remember